lsp/nls/src/position.rs

```diff
@@ -0,0 +1,162 @@
+use codespan::ByteIndex;
+use nickel_lang_core::{
+    position::TermPos,
+    term::{RichTerm, SharedTerm, Traverse, TraverseControl},
+};
+use range_map::{Range, RangeMap};
+
+// A term that uses pointer equality to implement Eq.
+#[derive(Clone, Debug)]
+struct RichTermPtr(RichTerm);
+
+impl PartialEq for RichTermPtr {
+    fn eq(&self, other: &Self) -> bool {
+        self.0.pos == other.0.pos && SharedTerm::ptr_eq(&self.0.term, &other.0.term)
+    }
+}
+
+impl Eq for RichTermPtr {}
+
+/// A lookup data structure, for looking up the term at a given position.
+///
+/// Overlapping positions are resolved in favor of the smaller one; i.e., lookups return the
+/// most specific term for a given position.
+#[derive(Clone, Debug)]
+pub struct PositionLookup {
+    table: RangeMap<u32, RichTermPtr>,
+}
+
+impl PositionLookup {
+    /// Create a position lookup table for looking up subterms of `rt` based on their positions.
+    pub fn new(rt: &RichTerm) -> Self {
+        let mut all_ranges = Vec::new();
+
+        rt.traverse_ref(&mut |term: &RichTerm| {
+            if let TermPos::Original(pos) = &term.pos {
+                all_ranges.push((pos.start.0, pos.end.0, term.clone()));
+            }
+            TraverseControl::<()>::Continue
+        });
+
+        // We rely on the invariant that if two ranges overlap then one is contained in the
+        // other. That is, the ranges form a tree. This sort order corresponds to pre-order
+        // traversal of that tree.
+        //
+        // (It would be nice if we could build the lookup table directly from the traversal
+        // in traverse_ref, but (1) the state tracking is fiddly and (2) the traversal order
+        // in traverse_ref can be weird; for example, a `Term::Annotated` sometimes contains
+        // children whose position is outside the position of the parent.)
```

Comment:
Yes, maybe this is just a bug? Concretely, it happens in

```
let x | { _ : { foo : Number | default = 1 } } = {} in x.PATH.y
```

There is an Annotated term whose position is just the two characters in `{}`, but traversing it also visits the `{ foo : Number | default = 1 }` term.

Reply:
Ah, this is for annotations put on a let-binding. I remember now. It's a bit harder because this code is desugared at some point to `{} | { ..contract.. }`, but it's not yet in this form while setting positions.

One solution could be to take everything from `|` to the closing `}` to be part of the annotated term, but that includes the equal sign, which is a bit unsettling. But this might be a pragmatic solution.

Another solution would be to follow more closely the source and store this annotation at the level of the let-binding inside the AST, as we do for annotation attached to record fields. This might duplicate a few match cases, but this sounds like the good thing to do to me. This is a refactoring PR on its own though.